### PR TITLE
start writing Serial wrapper

### PR DESCRIPTION
**Pull request by DhruvaG2000**
Need help with println

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

### CODE REVIEW

**Comment by DhruvaG2000** [comment]
will try something like ``#define Serial.println(f_, ...) printf((f_), ##__VA_ARGS__)``
reference: https://stackoverflow.com/a/20639864

File: cores/arduino/zephyrSerial.cpp [hunk 1]
```diff
@@ -11,12 +11,12 @@ size_t arduino::ZephyrSerial::print(char ch){
     return 0;
 }
 size_t arduino::ZephyrSerial::println(const char* ptr) {
-    printf("%s\n", ptr);
+    printf("\n%s\n", ptr);
```
Comment (szczys) [inline cores/arduino/zephyrSerial.cpp]: `println` should not have a newline at the beginning. This way you can do printf-like things on the same line:

```
Serial.print("Here is a number: ");
Serial.println(someNumber);
```

File: cores/arduino/zephyrSerial.cpp [hunk 2]
```diff
@@ -1,37 +1,28 @@
...
-size_t arduino::ZephyrSerial::println(unsigned int, int){}
-size_t arduino::ZephyrSerial::println(long, int){}
-size_t arduino::ZephyrSerial::println(unsigned long, int){}
-size_t arduino::ZephyrSerial::println(long long, int){}
-size_t arduino::ZephyrSerial::println(unsigned long long, int){}
-size_t arduino::ZephyrSerial::println(double, int ){}
-size_t arduino::ZephyrSerial::println(const Printable&){}
-size_t arduino::ZephyrSerial::println(void){}
\ No newline at end of file
+size_t arduino::ZephyrSerial::println(const char* ptr) {
```
Comment (szczys) [inline cores/arduino/zephyrSerial.cpp]: There should be a `print(const char* ptr)` version of this function.

File: cores/arduino/zephyrSerial.cpp [hunk 3]
```diff
@@ -1,37 +1,28 @@
...
+#include <zephyrSerial.h>
+#include <Arduino.h>
+#include <zephyr/sys/printk.h>
 
-size_t arduino::ZephyrSerial::print(const __FlashStringHelper *) {}
-size_t arduino::ZephyrSerial::print(const String &){}
-size_t arduino::ZephyrSerial::print(const char[]){
-    printk("\n Called print \n");
+size_t arduino::ZephyrSerial::begin(unsigned long int baudrate){
     return 0;
```
Comment (szczys) [inline cores/arduino/zephyrSerial.cpp:6]: Functions should return the size_t for the number of bytes written.

**Review by szczys — COMMENTED** [review]
@DhruvaG2000 please review my comments which I addressed with commit dda472a2e97ebef15dfe9e10602e427828303795

I think we should also implement [the rest of the print functions](https://www.arduino.cc/reference/en/language/functions/communication/serial/print/). Do you want to get this PR merged and add the others in a future PR?
* Serial.print(int)
* Serial.print(int, OCT | BIN | HEX | DEC)
* Serial.print(float)
* Serial.print(float, num_decimal_places)

**Comment by DhruvaG2000** [comment]
@szczys Thanks alot for the cleaner implementations, however I am unable to build, am I missing something?
Error:
```sh
[163/173] Linking CXX executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/zephyr_pre0.map
: && ccache /home/dhruva/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-g++   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/arm/common/cortex_m/libsoc__arm__common__cortex_m.a  zephyr/soc/arm/nordic_nrf/nrf52/libsoc__arm__nordic_nrf__nrf52.a  zephyr/boards/arm/arduino_nano_33_ble/libboards__arm__arduino_nano_33_ble.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  modules/hal_nordic/nrfx/libmodules__hal_nordic__nrfx.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/dhruva/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/thumb/v7e-m/nofp"  -L/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcpu=cortex-m4  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr && /usr/bin/cmake -E echo
/home/dhruva/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/libzephyr.a(zephyrSerial.cpp.obj): in function `arduino::ZephyrSerial::println()':
/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/cores/arduino/zephyrSerial.cpp:36: undefined reference to `putchar'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build
[ble: exit 1][ble: elapsed 6.014s (CPU 379.8%)] west build -p -b arduino_nano_33_ble samples/hello_arduino   -DZEPHYR_EXTRA_MODULES=/home/dhruva/zephyrproject/modules/lib/gso
```

**Comment by szczys** [comment]
I made a comment on your latest commit. Once you address that issue, this PR LGTM!

**Review by szczys — APPROVED** [review]

File: cores/arduino/zephyrSerial.cpp [hunk 4]
```diff
@@ -6,23 +6,35 @@ size_t arduino::ZephyrSerial::begin(unsigned long int baudrate){
...
 }
 
+size_t arduino::ZephyrSerial::print_char(char ch, bool lf){
+    printk(lf ? "%c\n" : "%c", ch);
+    return lf ? 2 : 1;
+}
+
+size_t arduino::ZephyrSerial::print_str(const char* ptr, bool lf) {
+    printf(lf ? "%s\n" : "%s", ptr);
+    return lf ? strlen(ptr)+1 : strlen(ptr);
```
Comment (szczys) [inline cores/arduino/zephyrSerial.cpp:16]: I have made an error here. This should be `printk` not `printf`.

I wonder if this was also part of the issue with your recent build test failing?

File: cores/arduino/zephyrSerial.cpp [hunk 5]
```diff
@@ -6,23 +6,35 @@ size_t arduino::ZephyrSerial::begin(unsigned long int baudrate){
     return 0;
 }
 
+size_t arduino::ZephyrSerial::print_char(char ch, bool lf){
+    printk(lf ? "%c\n" : "%c", ch);
+    return lf ? 2 : 1;
+}
+
+size_t arduino::ZephyrSerial::print_str(const char* ptr, bool lf) {
+    printf(lf ? "%s\n" : "%s", ptr);
```
Comment (szczys) [inline cores/arduino/zephyrSerial.cpp:15]:
```suggestion
   printk(lf ? "%s\n" : "%s", ptr);
```

I have made an error here. This should be `printk`, not `printf`.

File: cores/arduino/zephyrSerial.cpp [hunk 6]
```diff
@@ -33,7 +33,7 @@ size_t arduino::ZephyrSerial::println(const char* ptr){
 }
 
 size_t arduino::ZephyrSerial::println(void){
-    printf("\n");
+    print_char('\n', true);
```
Comment (szczys) [inline cores/arduino/zephyrSerial.cpp:36]:
```suggestion
    printk("\n");
```

I suspect the issue was with calling `printf` when we should call `printk`.

I would prefer to make the direct call here as I think it's more readable than calling `print_char` to do the same thing. Also, this commit as it's currently written would print the newline character twice (once for the `\n` character and again for the `true` parameter.

**Review by szczys — CHANGES_REQUESTED** [review]

File: cores/arduino/zephyrSerial.cpp (hunk 6 above)
Comment (DhruvaG2000) [reply cores/arduino/zephyrSerial.cpp:36]: Build succeeds with the above suggested changes:

```
└─❯ west build -p -b arduino_nano_33_ble samples/hello_arduino   -DZEPHYR_EXTRA_MODULES=/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core
-- west build: making build dir /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/samples/hello_arduino
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.4") found components: Interpreter
-- Cache files will be written to: /home/dhruva/.cache/zephyr
-- Zephyr version: 3.1.99 (/home/dhruva/zephyrproject/zephyr)
-- Found west (found suitable version "0.13.1", minimum required is "0.7.1")
-- Board: arduino_nano_33_ble
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.14.2 (/home/dhruva/zephyr-sdk-0.14.2)
-- Found dtc: /home/dhruva/zephyr-sdk-0.14.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6")
-- Found toolchain: zephyr 0.14.2 (/home/dhruva/zephyr-sdk-0.14.2)
-- Found BOARD.dts: /home/dhruva/zephyrproject/zephyr/boards/arm/arduino_nano_33_ble/arduino_nano_33_ble.dts
-- Found devicetree overlay: /home/dhruva/zephyrproject/zephyr/../modules/lib/Arduino-Zephyr-API/variants/arduino_nano_33_ble/arduino_nano_33_ble.overlay
-- Generated zephyr.dts: /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/dts.cmake
Parsing /home/dhruva/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/dhruva/zephyrproject/zephyr/boards/arm/arduino_nano_33_ble/arduino_nano_33_ble_defconfig'
Merged configuration '/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/samples/hello_arduino/prj.conf'
Configuration saved to '/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/.config'
Kconfig header saved to '/home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/dhruva/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dhruva/zephyrproject/modules/lib/gsoc-2022-arduino-core/build
-- west build: building application
[1/174] Preparing syscall dependency handling

[3/174] Generating include/generated/version.h
-- Zephyr version: 3.1.99 (/home/dhruva/zephyrproject/zephyr), build: zephyr-v3.1.0-3166-g0b6f4c7537ba
[164/174] Linking CXX executable zephyr/zephyr_pre0.elf

[168/174] Linking CXX executable zephyr/zephyr_pre1.elf

[174/174] Linking CXX executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       19428 B       928 KB      2.04%
            SRAM:        4352 B       256 KB      1.66%
        IDT_LIST:          0 GB         2 KB      0.00%

```

**Review by szczys — APPROVED** [review]

**Comment by DhruvaG2000** [comment]
> Do you want to get this PR merged and add the others in a future PR?

I will work over it this weekend, till then I think we can merge this. Will create a new PR if the requirement for the remaining arises.

File: samples/hello_arduino/src/app.cpp [hunk 7]
```diff
@@ -3,9 +3,27 @@
...
+  size_t ret2;
+  ret1 = Serial.print(c);
+  ret2 = Serial.println("Hello, World!");
+  printk("Sizes: %d %d\n", ret1, ret2);
+  Serial.println();
+  ret1 = Serial.print("My letter is: ");
+  ret2 = Serial.println(c);
+  printk("Sizes: %d %d\n", ret1, ret2);
+  Serial.println();
+  char myString[] = "Will it print?";
```
Comment (DhruvaG2000) [inline samples/hello_arduino/src/app.cpp:25]: Hmm it seems like we missed cleaning this bit up back then and merged it. Never really had to go back to hello arduino sample so didn't notice this until now. Will create a new PR to clean all this up.

**Pull request merged**